### DISCUSSION TITLE
Incremental or online representation learning

### DISCUSSION THREAD

**Discussion by bernddude** [lmcinnes/umap#843] [Ideas]
Is there any way how one could use umap in an online or incremental representation learning context ?
It would seem that this could be a very attractive extension . I was wondering to keep the dat points which provide more information and continuously discard less relevant or duplicated data points. However short of writing she external code at uses essentially the density in the final space i couldn't see an obvious way how to do this ?
The solution i just sketched out does not seem to be very elegant

**Comment by jc-healy** [comment]
It probably depends on what the desired properties of such an incremental
learning function might be.  For the moment I'll ignore efficiencies in
retaining unique or important data and focus on the core of an incremental
algorithm.

There are two quick ways to play such games.  The easiest is probably to
make use of parametric-UMAP (
https://umap-learn.readthedocs.io/en/latest/parametric_umap.html) which,
under the hood, trains a neural net for performing the embedding.  You
could then update the model with new training data in a (probably batch)
streaming manner.

If you mostly care about ensuring continuity between your sequential
embeddings then you could maintain a sliding window of data and look at a
sequence of pairwise aligned UMAP runs (
https://umap-learn.readthedocs.io/en/latest/aligned_umap_basic_usage.html).

Cheers,
John

**Reply by bernddude** [reply]
Thank you so much for the response John, those are really interesting ideas. I really do appreciate the suggestions. I did indeed wonder a bit about the 1st option a little bit. However to be blunt i wasn't sure that i understand enough of the details how the parametric-UMAP approach is implemented to judge it on my own. I'll definitely give this a try. just a question since i am not that familiar with the forum here. Would it help if i post/share what ever i find in this post ?

**Comment by jc-healy** [comment]
My general philosophy is to err on the side of sharing.  If you've
encountered a problem it's likely that other folks have as well.  As such
having (even partial) solutions or simply future problems that might come
up along the way in a searchable public forum can be quite helpful to
others.

**Reply by bernddude** [reply]
Ok , thank you John

**Comment by bernddude** [comment]
Hello John, i looked a bit more into your suggestions and at the actual implementation of the parametric uMAP module.
if i see this right it seems that the way to do this might be to modify the implementation of parametric uMAP in a way that one could provide a TF model as command line parameter (one with per-trained weights from a prior fit run). Also likely add a get function for the model and to modify the fit function in a way that if a pre-trained model is provided the TF initialization is skipped and the provided model is used instead ?
i that what you had in mind too ?
I didn't quite see how i could use the existing package w/o modifications 

thanks again Bernd
